product-service/src/main/java/org/arpha/mapper/helper/ProductMapperHelper.java

```diff
@@ -0,0 +1,64 @@
+package org.arpha.mapper.helper;
+
+import lombok.RequiredArgsConstructor;
+import org.arpha.entity.Category;
+import org.arpha.entity.Genre;
+import org.arpha.mapper.CategoryMapper;
+import org.arpha.mapper.GenreMapper;
+import org.arpha.repository.CategoryRepository;
+import org.arpha.repository.GenreRepository;
+import org.mapstruct.Named;
+import org.springframework.stereotype.Component;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+@RequiredArgsConstructor
+public class ProductMapperHelper {
+
+    private final CategoryMapper categoryMapper;
+    private final GenreMapper genreMapper;
+    private final CategoryRepository categoryRepository;
+    private final GenreRepository genreRepository;
+
+    @Named("toStringGenres")
+    public Set<String> toStringGenres(Set<Genre> genres) {
+        return genres.stream().map(Genre::getName).collect(Collectors.toSet());
+    }
+
+    @Named("toStringCategories")
+    public Set<String> toStringCategories(Set<Category> categories) {
+        return categories.stream().map(Category::getName).collect(Collectors.toSet());
+    }
+
+    @Named("toGenres")
+    public Set<Genre> toGenres(Set<String> genres) {
+        Set<Genre> genreSet = new HashSet<>();
+        for (String genreName : genres) {
+            if(genreRepository.existsByName(genreName)) {
+                genreSet.add(genreRepository.findByName(genreName));
+            } else {
+                genreSet.add(genreRepository.save(genreMapper.toGenre(genreName)));
```

Comment:
Same as above

Reply:
I'll have circular dependency because I need abstract-product-service include Genre entity as return from product-service module, and product-service uses abstract-product-service, so I'll need to get response as some Dto and then again map back to jpa entity. Seems pointless a bit. I'd rather have logic in that case in helper. If you know better solution write pls